### DISCUSSION TITLE
Want to render code or diff in diagnostic message

### DISCUSSION THREAD

**Discussion by Chhinna** [microsoft/vscode-discussions#1094] [Extension Development QnA] [answered]
I want to render diagnostics, in my vs code. But instead of just string message, I want to render diff or code. How can I do it? I was exploring `new vscode.Diagnostic` as well as `new vscode.DiagnosticRelatedInformation`.

You can refer to this gif, where a -, + diff is rendered. https://raw.githubusercontent.com/sourcery-ai/sourcery-vscode/main/assets/Sourcery_VS_Code_Refactoring.gif 

Any help would be appreciated.

**Comment by abrilurena** [comment, marked as answer]
Hi, @Chhinna! I found this, please let me know if it helps you: The built-in VS Code diagnostics do not directly support rendering code or diffs within the diagnostic messages. However, there are a few workarounds you can consider:

1. Custom Formatting in Diagnostic Messages:
- While the default vscode.Diagnostic class does not directly support rich formatting (such as rendering diffs), you can still include additional information in the diagnostic message.
- You could create a custom format for your diagnostic messages, including the relevant code snippets or diffs. 
2. Markdown in Diagnostic Messages:
- Although not officially supported, some extensions use Markdown formatting in diagnostic messages. You can experiment with this approach.
- For instance, you could try including Markdown-formatted code blocks or inline code snippets in your diagnostic messages.
3. Custom Extension or Language Server:
- If your use case requires more advanced features, consider creating a custom extension or a language server.
- A custom extension can provide a more tailored solution for rendering diffs or code snippets in diagnostics.
- Language servers (such as those used by TypeScript, Python, etc.) often have more flexibility in providing detailed diagnostic information.
4. Third-Party Extensions:
- Explore existing extensions that might already offer the functionality you need.
- Some extensions, like Sourcery, might provide custom diagnostic rendering capabilities. You can check if they expose APIs or settings for this purpose.

I hope this helps, and feel free to explore the options mentioned above to achieve your desired outcome! 😊
